### ISSUE TITLE
Consider default export convenience

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#304]
6to5 implemented a feature where ES6 modules with just a single default export have the default export treated as the whole module when loading in CommonJS.

We could consider this feature for SystemJS too. It would be breaking so the time to make it would be along with all this other stuff.

Feedback would be great to hear here.

Basically it would allow:

es6.js
```javascript
export default function() {
}
```

cjs.js
```javascript
var fn = require('es6');
// instead of
var fn = require('es6').default;
```

**Comment by karlbohlmark** [comment]
This would be very much appreciated!

**Comment by trusktr** [comment]
Just took a peek here: https://github.com/systemjs/systemjs/blob/master/lib/extension-core.js#L11 That description sounds similar to ths idea.

**Comment by trusktr** [comment]
Maybe __useDefault isn't being set to true in my case because I also have named exports? I'll test in 7 hours.

**Comment by trusktr** [comment]
Ok, maybe not 7 hours, but having just a default export only didn't work.

**Comment by guybedford** [comment]
This is a separate feature to `__useDefault`, but a related concept yes.

**Comment by tauren** [comment]
:+1:

**Comment by highsource** [comment]
Here's an example for this:

* https://rawgit.com/highsource/javascript-module-loaders-comparison/master/EcmaScript6/SystemJS/index.html

What I do in Main.js is export a single object as default:

```javascript
var greeting = new Greets.Greeting("Hello", "World", new Greets.ConsoleMessenger());

export default greeting;
```

I then expect this very object loaded in `then`:

```javascript
System.import('Main').then(function(greeting) {
	greeting.greet();
});
````

But instead I have to do this:

```javascript
System.import('Main').then(function(greeting) {
	greeting.default.greet();
});
````

Which is a bit awkward.

**Comment by MajorBreakfast** [comment]
I like this idea.

I don't know why someone would want to write a CommonJS module that depends on an es6 module, though. This kinda requires using SystemJS and then going all es6 is the wiser choice. However, I still like it. It seems sensible.

@highsource You should really include in your comment that you're talking about the `__useDefault` feature. What you're describing is basically when you'll want to include `export __useDefault true` in your module to avoid the `greeting.default.greet()` mess.

**Comment by highsource** [comment]
@MajorBreakfast I'm not quite sure about your comment.

First, I'm not writing a CommonJS module that depends on es6. At least this is not my intention. I just want to call `greeting.greet()` in my web page (like [here](view-source:https://rawgit.com/highsource/javascript-module-loaders-comparison/master/Globals/index.html) in the globals example.

Next, I'm not quite sure what you mean with `__useDefault`. From what I read [here](http://www.2ality.com/2014/09/es6-modules-final.html), this:

```javascript
export default greeting;
```

Should actually "just" export the `greeting`. So I'm not sure why should I need `__useDefault` here.

But I'm very new to this so maybe getting things wrong.

**Comment by MajorBreakfast** [comment]
You should do:
``` JavaScript
var greeting = new Greets.Greeting("Hello", "World", new Greets.ConsoleMessenger())
export default greeting
export __useDefault true
```

which makes this possible:
``` JavaScript
System.import('Main')
  .then(function(greeting) { greeting.greet() })
```

**Comment by highsource** [comment]
Ok, I see. Thank you for the reply.

This:

```javascript
export default greeting
export __useDefault true
```

did not work but this:

```javascript
export default greeting;
export var __useDefault = true;
```
did.

@MajorBreakfast Do you happen to know where is this `__useDefault` thing is specified? Could not find a spec reference for it...

Thank you for the `__useDefault` hint.

**Comment by MajorBreakfast** [comment]
@highsource Maybe that's better answered by @guybedford but here's what I can gather:
- It's implemented in SystemJS.
- It seems to be a private SystemJS feature, because I can't find documentation for it in the readme
- Judging by the amount of github issues it has also it's edge cases.
- It seems that the `System.import`, AMD and the register format already respect `__useDefault`

@guybedford It'd be nice if you could correct these statements. I'm curious :)

**Comment by guybedford** [comment]
The only thing we do around any of this at the moment is ensure that `System.import` provides the `default` value of an export, just by patching the `System.import` function here - https://github.com/systemjs/systemjs/blob/master/lib/extension-core.js#L25.

**Comment by highsource** [comment]
@guybedford 

Yes, I've found this. I'd just want to make an ES6-compliant "default" export and I though `export default greeting;` would be enough.

I'm OK with the `__useDefault` workaround at the moment, no big deal. But when ES6 will be finalized, would be nice to have SystemJS compliant to the spec.

**Comment by MajorBreakfast** [comment]
@guybedford But there are other places with `useDefault` logic which I didn't expect. I don't quite know what those do:
- [register extension line 304](https://github.com/systemjs/systemjs/blob/61ac01e3f82c5c5d5222c6ad3badb9f59e9d03ab/lib/extension-register.js#L304), [353](https://github.com/systemjs/systemjs/blob/61ac01e3f82c5c5d5222c6ad3badb9f59e9d03ab/lib/extension-register.js#353), [561](https://github.com/systemjs/systemjs/blob/61ac01e3f82c5c5d5222c6ad3badb9f59e9d03ab/lib/extension-register.js#561)
- [amd extension line 75](https://github.com/systemjs/systemjs/blob/99e547700201a7c55c262907238048ab904a5dd4/lib/extension-amd.js#L75)

**Comment by guybedford** [comment]
So the logic is:

* AMD, CommonJS and globals (dynamic modules) are always defined as a `default` export with a `__useDefault` (private API, not specified or will ever be specified, but spec-compatible as a private API) (register extension lines 304, 561)
* When loading a dependency of CommonJS or AMD, we take the value as the default export (register extension line 353, AMD extension 75).

**Comment by highsource** [comment]
@guybedford Just for me to understand. Why not take the `default` value as `default` value in case ot ES6? Or should it be addressed in the 6-5 transpiller?

**Comment by guybedford** [comment]
@highsource yes we can consider that, and that is what this issue is tracking. 

I'm hesitant though, as we should alter spec behaviour carefully. 

Specifically, it is not 100% certain if the patching of System.import will be possible. For example, a scoped `System.import` will be automatically created in the module body for relative dynamic requires - `import { import } from this module` or something like that. There may be no way to alter this scoped import like we do in https://github.com/systemjs/systemjs/blob/master/lib/extension-core.js#L25, in which case we can't break from spec behaviour.

**Comment by MajorBreakfast** [comment]
@highsource ~~6to5~~ It's "babel" now :P haha

- I think the default export convenience for `System.import()` in its form today is already problematic since it relies on behavior not in the spec. In due time users will create es6 modules which will need SystemJS for nothing else but that feature. That's not good.
- The `__useDefault` of modules loaded inside a non-es6 module can trigger whatever behavior we like since SystemJS will be there anyway. But I have a bad feeling if its automatic and not explicit, though. It reminds me of the old behavior of `RSVP.denodeify()` (e.g. current version used in jspm) which behaved differently depending on whether the callback had two parameters (`value` and `error`) or more. If it had two it transformed to `.then(function(value) {})` if it had more it transformed to `.then(function([value1, value2, ...]) {})`. [Not anymore](https://github.com/tildeio/rsvp.js/blob/master/lib/rsvp/node.js#L97-L98). In other words not explicit means that a library author can break stuff with seemingly harmless changes.

Those are my thoughts on that. I'd prefer if there was a spec compliant way. Some way for library authors to specify `__useDefault` or something much simpler like `System.importDefault()`.

**Comment by guybedford** [comment]
@MajorBreakfast this is why `__useDefault` is not documented - it is not a public API feature of SystemJS and is not something users should use. It simply a flag for making internal behaviour work.

Having `System.import` not follow it was a consideration, but there is no way we could provide an API to users that didn't give them what they expected for CommonJS imports.

**Comment by guybedford** [comment]
After having a discussion with @matthewp we're reconsidering if this should be supported.

It is really strange that adding an extra export creates a breaking change in a module.

**Comment by MajorBreakfast** [comment]
I'm fairly certain that there is not right answer here :) Let's face it: The ES format has a feature that CommonJS has not.

I think we can agree on:
- `require('my-module').default` is awful -> Pro
- Adding another named export causes bugs if this is enabled -> Contra

My suggestion: Add it anyway because @guybedford mentioned that babel already has this feature.

It's not optimal either way. Besides, I don't think that many people code CommonJS modules that depend on ES modules, so this is an edge case. To opt-out users can simply add a random named export in advance.

**Comment by matthewp** [comment]
Consider you are a library author that doesn't use SystemJS. Your code breaks for SystemJS users because you add a new export? What do you do in this scenario, major version bump your library just because of a weird SystemJS "feature"?  I don't think adding a fake export to every file is a solution.

Breaking people's code is a real serious issue. People who encounter this bug are not going to understand the difference between default and named exports, they are just going to wonder why their code suddenly broken because they did a `jspm update` and got a patch from some lib.

**Comment by guybedford** [comment]
It's worth noting that in Babel it sort of makes sense because it is just when transpiling to CommonJS for consumption in npm / Node.

One doesn't have to expect that the transpiled code into CommonJS via Babel has exactly the same interface as raw ES6.

It is an edge case here, and most users will write ES6 which is getting conveniences for named exports from CommonJS.

So I tend to agree with the arguments against.

**Comment by MajorBreakfast** [comment]
> It's worth noting that in Babel it sort of makes sense because it is just when transpiling to CommonJS for consumption in npm / Node.

That is a very good point.

**Comment by guybedford** [comment]
This is no longer being implemented for the reasons discussed.
